File: packages/python/src/Seguranca/Ataques/AtaqueAvalancheChave.py
"""O AtaqueAvalanche mede a difusão de 1 bit do IV; esse faz o mesmo com a
CHAVE. É o teste mais direto de "a chave está de fato sendo misturada por
inteiro": mudar 1 bit da chave deveria mudar ~50% dos bits do keystream.
Números baixos indicam que parte da chave tem pouca influência - o que
reduziria o espaço de busca de um atacante.
"""

from ..ResultadoAtaque import ResultadoAtaque
from ..SkipAtaqueException import SkipAtaqueException
from ..Util import bits_diferentes, random_bytes, random_int


class AtaqueAvalancheChave:
    def __init__(
        self,
        amostras: int = 3000,
        tamanho_bloco: int = 32,
        limite_media: float = 0.45,
        limite_pior_caso: float = 0.30,
    ):
        self.amostras = amostras
        self.tamanho_bloco = tamanho_bloco
        self.limite_media = limite_media
        self.limite_pior_caso = limite_pior_caso

    def nome(self) -> str:
        return "Efeito avalanche da chave"

    def executar(self, alvo) -> ResultadoAtaque:
        chave_base = alvo.chave_de_teste()
        len_chave = len(chave_base)
        iv = random_bytes(alvo.tamanho_iv())

        primeiro = alvo.gerar_keystream_bruto(chave_base.encode(), iv, "enc", self.tamanho_bloco)
        if primeiro is None:
            raise SkipAtaqueException("Alvo não expõe gerar_keystream_bruto.")

        valores = []
        for t in range(self.amostras):
            chave = bytearray(chave_base.encode())
            pos = random_int(0, len_chave - 1)
            chave[pos] ^= 1 << random_int(0, 7)

            ks1 = alvo.gerar_keystream_bruto(chave_base.encode(), iv, "enc", self.tamanho_bloco)
            ks2 = alvo.gerar_keystream_bruto(bytes(chave), iv, "enc", self.tamanho_bloco)

            valores.append(bits_diferentes(ks1, ks2) / (self.tamanho_bloco * 8))

        valores = sorted(valores)
        n = len(valores)
        media = sum(valores) / n
        pior_caso = valores[0]

        vulneravel = media < self.limite_media or pior_caso < self.limite_pior_caso

        return ResultadoAtaque(
            self.nome(),
            vulneravel,
            "alta" if vulneravel else "info",
            (
                f"média={media * 100:.1f}%, pior caso={pior_caso * 100:.1f}% "
                f"(limites: média>={self.limite_media * 100:.0f}%, pior>={self.limite_pior_caso * 100:.0f}%)"
            ),
            {"media": media, "pior_caso": pior_caso},
        )

File: packages/typescript/src/Seguranca/Ataques/AtaqueAvalancheChave.ts
import { ResultadoAtaque } from '../ResultadoAtaque.ts';
import { SkipAtaqueException } from '../SkipAtaqueException.ts';
import { randomInt, randomBytes, bitsDiferentes } from '../Util.ts';
import type { AlvoCriptografico } from '../AlvoCriptografico.ts';
import type { AtaqueInterface } from '../AtaqueInterface.ts';

/**
 * O AtaqueAvalanche mede a difusão de 1 bit do IV; esse faz o mesmo com a
 * CHAVE. É o teste mais direto de "a chave está de fato sendo misturada por
 * inteiro": mudar 1 bit da chave deveria mudar ~50% dos bits do keystream.
 * Números baixos indicam que parte da chave tem pouca influência - o que
 * reduziria o espaço de busca de um atacante.
 */
export class AtaqueAvalancheChave implements AtaqueInterface {
  private readonly amostras: number;
  private readonly tamanhoBloco: number;
  private readonly limiteMedia: number;
  private readonly limitePiorCaso: number;

  constructor({
    amostras = 3000,
    tamanhoBloco = 32,
    limiteMedia = 0.45,
    limitePiorCaso = 0.3,
  }: {
    amostras?: number;
    tamanhoBloco?: number;
    limiteMedia?: number;
    limitePiorCaso?: number;
  } = {}) {
    this.amostras = amostras;
    this.tamanhoBloco = tamanhoBloco;
    this.limiteMedia = limiteMedia;
    this.limitePiorCaso = limitePiorCaso;
  }

  nome(): string {
    return 'Efeito avalanche da chave';
  }

  executar(alvo: AlvoCriptografico): ResultadoAtaque {
    const chaveBase = Buffer.from(alvo.chaveDeTeste(), 'binary');
    const lenChave = chaveBase.length;
    const iv = randomBytes(alvo.tamanhoIv());

    const primeiro = alvo.gerarKeystreamBruto(chaveBase, iv, 'enc', this.tamanhoBloco);
    if (primeiro === null) {
      throw new SkipAtaqueException('Alvo não expõe gerarKeystreamBruto.');
    }

    const valores: number[] = [];
    for (let t = 0; t < this.amostras; t++) {
      const chave = Buffer.from(chaveBase);
      const pos = randomInt(0, lenChave - 1);
      chave[pos] = chave[pos] ^ (1 << randomInt(0, 7));

      const ks1 = alvo.gerarKeystreamBruto(chaveBase, iv, 'enc', this.tamanhoBloco);
      const ks2 = alvo.gerarKeystreamBruto(chave, iv, 'enc', this.tamanhoBloco);

      valores.push(bitsDiferentes(ks1, ks2) / (this.tamanhoBloco * 8));
    }

    valores.sort((a, b) => a - b);
    const n = valores.length;
    const media = valores.reduce((acc, v) => acc + v, 0) / n;
    const piorCaso = valores[0];

    const vulneravel = media < this.limiteMedia || piorCaso < this.limitePiorCaso;

    return new ResultadoAtaque(
      this.nome(),
      vulneravel,
      vulneravel ? 'alta' : 'info',
      `média=${(media * 100).toFixed(1)}%, pior caso=${(piorCaso * 100).toFixed(1)}% (limites: média>=${(this.limiteMedia * 100).toFixed(0)}%, pior>=${(this.limitePiorCaso * 100).toFixed(0)}%)`,
      { media: media, pior_caso: piorCaso },
    );
  }
}
